Bottom: dab112e2a740f985dd1d4e44f6f8bdee9e3c9733
Top:    b0dc57f2e858797d4d533d7b73ec9179eca0551d
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-06-14 11:38:18 -0500

Refresh of common-configure-docker-daemon

---
diff --git a/ansible-playbook/group_vars/docker-hosts/docker.yml b/ansible-playbook/group_vars/docker-hosts/docker.yml
index ff71231f..f9153cdc 100644
--- a/ansible-playbook/group_vars/docker-hosts/docker.yml
+++ b/ansible-playbook/group_vars/docker-hosts/docker.yml
@@ -24,3 +24,5 @@ docker_network_frontend: 'frontend'
 
 # docker force build images
 docker_image_force_build: true
+
+docker_storage_driver: overlay2
diff --git a/ansible-playbook/host_vars/testdaedalus.novatech-llc.com/docker.yml b/ansible-playbook/host_vars/testdaedalus.novatech-llc.com/docker.yml
new file mode 100644
index 00000000..df7bb7de
--- /dev/null
+++ b/ansible-playbook/host_vars/testdaedalus.novatech-llc.com/docker.yml
@@ -0,0 +1,3 @@
+---
+
+docker_storage_driver: devicemapper
diff --git a/roles/docker-common/handlers/main.yml b/roles/docker-common/handlers/main.yml
new file mode 100644
index 00000000..8c329de7
--- /dev/null
+++ b/roles/docker-common/handlers/main.yml
@@ -0,0 +1,8 @@
+---
+- name: Restart Docker Daemon
+  systemd:
+    name: docker.service
+    daemon_reload: yes
+    enabled: yes
+    state: restarted
+  listen: "restart docker service"
diff --git a/roles/docker-common/tasks/configure.yml b/roles/docker-common/tasks/configure.yml
index c1f016e1..34754342 100644
--- a/roles/docker-common/tasks/configure.yml
+++ b/roles/docker-common/tasks/configure.yml
@@ -1,25 +1,59 @@
 ---
 
-- name: Create Docker Configuration Directory
+- name: Create Docker service configuration directory
   file:
-    path: /etc/docker
+    path: /etc/systemd/system/docker.service.d
     state: directory
     owner: root
     group: root
+
+- name: Configure Docker Daemon (host options)
+  template: 
+    dest: /etc/systemd/system/docker.service.d/00-host.conf
+    src: service-host.conf.j2
+    owner: root
+    group: root
+    mode: 'u=rw,g=r,o=r'
+  notify: "restart docker service"
+
+- name: Configure Docker Daemon (log driver options)
+  template: 
+    dest: /etc/systemd/system/docker.service.d/10-log.conf
+    src: service-log-json.conf.j2
+    owner: root
+    group: root
     mode: 'u=rw,g=r,o=r'
+  notify: "restart docker service"
+
+- assert:
+    that: docker_thinpool_lv is defined
+  when: docker_storage_driver == "devicemapper"
 
-- name: Configure Docker Daemon
+- name: Configure Docker Daemon (storage driver options)
   template: 
-    dest: /etc/docker/daemon.json
-    src: daemon.json.j2
+    dest: /etc/systemd/system/docker.service.d/20-storage.conf
+    src: service-storage-dm.conf.j2
     owner: root
     group: root
     mode: 'u=rw,g=r,o=r'
-  register: docker_json
+  when: docker_storage_driver == "devicemapper"
+  notify: "restart docker service"
 
-- name: Restart Docker Daemon
-  service:
-    name: docker
-    enabled: yes
-    state: reloaded
-  when: docker_json.changed
+- name: Configure Docker Daemon (storage driver options)
+  template: 
+    dest: /etc/systemd/system/docker.service.d/20-storage.conf
+    src: service-storage-ov2.conf.j2
+    owner: root
+    group: root
+    mode: 'u=rw,g=r,o=r'
+  when: docker_storage_driver == "overlay2"
+  notify: "restart docker service"
+
+- name: Configure Docker Daemon (service)
+  template: 
+    dest: /etc/systemd/system/docker.service.d/99-service.conf
+    src: service.conf.j2
+    owner: root
+    group: root
+    mode: 'u=rw,g=r,o=r'
+  notify: "restart docker service"
diff --git a/roles/docker-common/templates/service-host.conf.j2 b/roles/docker-common/templates/service-host.conf.j2
new file mode 100644
index 00000000..38c9f6e3
--- /dev/null
+++ b/roles/docker-common/templates/service-host.conf.j2
@@ -0,0 +1,5 @@
+[Service]
+Environment="HOST_OPTIONS=\
+--host=unix:///var/run/docker.sock \
+--host=tcp://{{ ansible_default_ipv4['address'] }}:2375 \
+"
diff --git a/roles/docker-common/templates/service-log-json.conf.j2 b/roles/docker-common/templates/service-log-json.conf.j2
new file mode 100644
index 00000000..a268ff76
--- /dev/null
+++ b/roles/docker-common/templates/service-log-json.conf.j2
@@ -0,0 +1,6 @@
+[Service]
+Environment="LOG_OPTIONS=\
+--log-driver=json-file \
+--log-opt max-size=10m \
+--log-opt max-file=3 \
+"
diff --git a/roles/docker-common/templates/service-storage-dm.conf.j2 b/roles/docker-common/templates/service-storage-dm.conf.j2
new file mode 100644
index 00000000..45532f6c
--- /dev/null
+++ b/roles/docker-common/templates/service-storage-dm.conf.j2
@@ -0,0 +1,8 @@
+[Service]
+Environment="STORAGE_OPTIONS=\
+--storage-driver devicemapper \
+--storage-opt dm.fs=xfs \
+--storage-opt dm.thinpooldev={{ docker_thinpool_lv.lv_dm_path }} \
+--storage-opt dm.use_deferred_removal=true \
+--storage-opt dm.use_deferred_deletion=true \
+"
diff --git a/roles/docker-common/templates/service-storage-ov2.conf.j2 b/roles/docker-common/templates/service-storage-ov2.conf.j2
new file mode 100644
index 00000000..3fa80a6e
--- /dev/null
+++ b/roles/docker-common/templates/service-storage-ov2.conf.j2
@@ -0,0 +1,4 @@
+[Service]
+Environment="STORAGE_OPTIONS=\
+--storage-driver overlay2 \
+"
diff --git a/roles/docker-common/templates/service.conf.j2 b/roles/docker-common/templates/service.conf.j2
new file mode 100644
index 00000000..7e552e00
--- /dev/null
+++ b/roles/docker-common/templates/service.conf.j2
@@ -0,0 +1,6 @@
+[Service]
+ExecStart=
+ExecStart=/usr/bin/dockerd \
+          $HOST_OPTIONS \
+          $LOG_OPTIONS \
+          $STORAGE_OPTIONS
